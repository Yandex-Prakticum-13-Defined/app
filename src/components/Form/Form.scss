.form {
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-weight: bold;
  font-size: 20px;

  &__wrapper {
    display: flex;
    flex-direction: column;
    margin: 5px auto;
    width: 80%;
    justify-content: flex-start;
  }

  &__input {
    display: flex;
    width: 80%;
    padding: 5px 25px;
    border-radius: 20px;
    border: none;
    color: inherit;
    outline: none;
    border-bottom: 3px solid black;
  }
}


.input {
  width: inherit;

  &__error {
    display:flex;
    width: 85%;
    color: red;
    margin-left: 10px;
    font-size: 14px;
    font-weight: normal;
  }
}
