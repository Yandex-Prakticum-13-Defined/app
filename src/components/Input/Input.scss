.input {
  padding: 5px 25px;
  border: none;
  border-bottom: 3px solid black;
  border-radius: 20px;

  &__error {
    color: red;
    margin-left: 10px;
    font-size: 13px;
    height: 33px;
  }
}
