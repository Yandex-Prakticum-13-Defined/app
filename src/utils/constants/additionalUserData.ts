export const additionalUserData = {
  second_name: 'Second',
  display_name: 'Display',
  phone: '+7-000-000-00-00'
};
