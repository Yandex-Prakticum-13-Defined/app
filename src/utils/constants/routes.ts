export enum ERoutes {
  'START' = '/',
  'REGISTER' = '/register',
  'LOGIN' = '/login',
  'LOGOUT' = '/logout',
  'GAME' = '/game',
  'LEADERBOARD' = '/leaderboard',
  'PROFILE' = '/profile',
  'FORUM' = '/forum',
  'ERROR_500' = '/500',
  'OFFLINE' = '/offline',
  'FALLBACK' = '*'
}

export enum EDBRoutes {
  'TOPICS' = '/topics',
  'MESSAGES' = '/messages',
  'TOPIC' = '/topic',
  'MESSAGE' = '/message'
}
