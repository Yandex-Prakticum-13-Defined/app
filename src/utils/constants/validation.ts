export const VALIDATION = {
  required: 'Поле не может быть пустым',
  name: {
    value: /^[A-ZА-ЯЁ]+[A_Za-zА-Яа-яЁё-]+/,
    message: 'Латиница или кириллица, первая буква заглавная, '
      + 'без пробелов и цифр, нет спецсимволов (допустим только дефис)'
  },
  login: {
    value: /[a-zA-Z0-9_-]*[a-zA-Z_-][a-zA-Z0-9_-]*/,
    message: 'Латиница, может содержать цифры, '
      + 'но не состоять из них, допустимы дефис и нижнее подчёркивание'
  },
  email: {
    value: /^[A-Za-z0-9,.-]{1,}[@]([A-Za-z0-9,.-]{1,}[.][A-Za-z]{1,}){1,2}$/,
    message: 'Латиница, может включать цифры и '
      + 'спецсимволы вроде дефиса, обязательна @ и точка после,'
      + ' но перед точкой обязательно должны быть буквы'
  },
  password: {
    value: /(?=.*[A-Z])(?=.*[0-9]).*/,
    message: 'От 8 до 40 символов, обязательно хотя бы одна заглавная буква и цифра'
  },
  minLength_8: {
    value: 8,
    message: 'Минимальное количество символов: 8'
  },
  minLength_3: {
    value: 3,
    message: 'Минимальное количество символов: 3'
  },
  maxLength_20: {
    value: 20,
    message: 'Максимальное количество символов: 20'
  }
};
