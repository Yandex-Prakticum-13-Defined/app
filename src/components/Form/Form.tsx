import React, { FC, FormEvent, ReactNode } from 'react';
import './Form.scss';
import { Link } from 'react-router-dom';
import Button, { IButton } from '../Button/Button';
import { ERoutes } from '../../App';

interface IForm {
  title: string;
  children: ReactNode;
  button: IButton;
  handleSubmit: (event: FormEvent<HTMLFormElement>) => void;
  linkTo: ERoutes;
  linkText: string;
}

const Form: FC<IForm> = ({
  title, children, button, handleSubmit, linkTo, linkText
}) => (
  <form className='form' onSubmit={handleSubmit}>
    <h1 className='form__title'>{title}</h1>
    {children}
    <Button type='submit' title={button.title} disabled={button.disabled}/>
    <Link className='form__link' to={linkTo}>{linkText}</Link>
  </form>
);

export default Form;
