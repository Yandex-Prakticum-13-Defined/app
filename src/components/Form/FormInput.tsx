import React, { memo } from 'react';
import './Form.scss';
import { useController } from 'react-hook-form';

interface IFormInput {
  control?: any;
  defaultValue?: string;
  name: string;
  rules?: any;
  type: string;
  placeholder: string;
  className?: string;
  onChange?: (e: any) => void;
  value?: string;
}

export const FormInput = memo(({
  control,
  defaultValue,
  name,
  rules,
  ...rest
}: IFormInput) => {
  const {
    field,
    formState: { errors }
  } = useController({
    name,
    control,
    rules,
    defaultValue: defaultValue ?? ''
  });

  return (
    <>
      <input {...field} {...rest} />
      {errors && <span className='input__error'>{errors[name]?.message}</span>}
    </>
  );
});
