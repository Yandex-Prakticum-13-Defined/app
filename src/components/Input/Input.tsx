import React from 'react';
import './Input.scss';
import { useController } from 'react-hook-form';

export interface IInput {
  name: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  control: any;
  value?: string;
  rules?: Record<string, unknown>;
  type: string;
  placeholder: string;
}

// eslint-disable-next-line react/display-name
export const Input = ({
  name,
  control,
  value,
  rules,
  ...rest
}: IInput) => {
  const {
    field,
    formState: { errors }
  } = useController({
    name,
    control,
    defaultValue: value ?? '',
    rules
  });

  return (
    <>
      {/* eslint-disable-next-line react/jsx-props-no-spreading */}
      <input className='input' {...field} {...rest}/>
      {errors && <span className='input__error'>{errors[name]?.message}</span>}
    </>
  );
};
