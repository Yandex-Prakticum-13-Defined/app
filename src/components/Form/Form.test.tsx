import React, { FC } from 'react';
import renderer from 'react-test-renderer';
import { useForm } from 'react-hook-form';
import { BrowserRouter } from 'react-router-dom';
import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { VALIDATION } from '../../utils/constants/validation';
import { Input } from '../Input/Input';
import { ERoutes } from '../../utils/constants/routes';
import Form from './Form';

const handleSubmit = jest.fn((event) => event.preventDefault());

const TestFormWrapper: FC = () => {
  const {
    formState: {
      isValid
    },
    control
  } = useForm({
    mode: 'onChange',
    defaultValues: {
      login: '',
      password: ''
    }
  });

  return (
    <BrowserRouter>
      <Form
        title='Заголовок'
        button={{
          type: 'submit',
          title: 'Кнопка',
          disabled: !isValid
        }}
        handleSubmit={handleSubmit}
        linkTo={ERoutes.REGISTER}
        linkText='Ссылка'
      >
        <Input
          name='login'
          type='text'
          placeholder='Логин'
          control={control}
          rules={{
            required: VALIDATION.required,
            minLength: VALIDATION.minLength_3,
            maxLength: VALIDATION.maxLength_20,
            pattern: VALIDATION.login
          }}
        />
        <Input
          name='password'
          type='password'
          placeholder='Пароль'
          control={control}
          rules={{
            required: VALIDATION.required,
            pattern: VALIDATION.password,
            minLength: VALIDATION.minLength_8
          }}
        />
      </Form>
    </BrowserRouter>
  );
};

describe('Тестируем Form.tsx', () => {
  it('проверяем снимок', () => {
    const inputComponent = renderer.create(
      <TestFormWrapper/>
    ).toJSON();
    expect(inputComponent).toMatchSnapshot();
  });

  it('содержит заголовок', () => {
    render(<TestFormWrapper/>);
    expect(screen.getByText('Заголовок')).toBeInTheDocument();
  });

  it('содержит дочерние инпуты', () => {
    render(<TestFormWrapper/>);
    expect(screen.getByPlaceholderText('Логин')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Пароль')).toBeInTheDocument();
  });

  it('содержит кнопку', () => {
    render(<TestFormWrapper/>);
    expect(screen.getByText('Кнопка')).toBeInTheDocument();
  });

  it('кнопка заблокирована при невалидных значениях', () => {
    render(<TestFormWrapper/>);
    expect(screen.getByText('Кнопка')).toBeDisabled();
  });

  it('кнопка активна при валидных значениях', async () => {
    render(<TestFormWrapper/>);
    const user = userEvent.setup();
    await act(async () => {
      await user.type(screen.getByPlaceholderText('Логин'), 'TestLogin');
    });
    await act(async () => {
      await user.type(screen.getByPlaceholderText('Пароль'), 'TestPassword1');
    });
    expect(screen.getByText('Кнопка')).toBeEnabled();
  });

  it('функция onSubmit вызывается при валидных значениях', async () => {
    render(<TestFormWrapper/>);
    const user = userEvent.setup();
    await act(async () => {
      await user.type(screen.getByPlaceholderText('Логин'), 'TestLogin');
    });
    await act(async () => {
      await user.type(screen.getByPlaceholderText('Пароль'), 'TestPassword1');
    });
    await act(async () => {
      await user.click(screen.getByText('Кнопка'));
    });
    expect(handleSubmit).toHaveBeenCalledTimes(1);
  });

  it('присутствует ссылка', () => {
    render(<TestFormWrapper/>);
    expect(screen.getByText('Ссылка')).toBeInTheDocument();
  });
});
