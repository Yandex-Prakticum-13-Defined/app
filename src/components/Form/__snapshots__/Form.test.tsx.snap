// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тестируем Form.tsx проверяем снимок 1`] = `
<form
  className="form"
  onSubmit={[MockFunction]}
>
  <h1
    className="form__title"
  >
    Заголовок
  </h1>
  <input
    className="input"
    name="login"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Логин"
    type="text"
    value=""
  />
  <span
    className="input__error"
  />
  <input
    className="input"
    name="password"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Пароль"
    type="password"
    value=""
  />
  <span
    className="input__error"
  />
  <button
    className="button"
    disabled={true}
    type="submit"
  >
    Кнопка
  </button>
  <a
    className="form__link"
    href="/register"
    onClick={[Function]}
  >
    Ссылка
  </a>
</form>
`;
